src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java

```diff
@@ -6469,6 +6469,244 @@ public final class TokenTypes {
     public static final int RECORD_PATTERN_COMPONENTS =
             JavaLanguageLexer.RECORD_PATTERN_COMPONENTS;
 
+    /**
+     * A string template opening delimiter. This element ({@code "}) appears
+     * at the beginning of a string template.
+     * <p>For example:</p>
+     * <pre>
+     *     String s = STR."Hello, \{name}!";
+     * </pre>
+     * <p>parses as:</p>
+     * <pre>
+     * `--VARIABLE_DEF -&gt; VARIABLE_DEF
+     *     |--MODIFIERS -&gt; MODIFIERS
+     *     |--TYPE -&gt; TYPE
+     *     |   `--IDENT -&gt; String
+     *     |--IDENT -&gt; s
+     *     `--ASSIGN -&gt; =
+     *         `--EXPR -&gt; EXPR
+     *              `--DOT -&gt; .
+     *                  |--IDENT -&gt; STR
+     *                  `--STRING_TEMPLATE_BEGIN -&gt; "
+     *                      |--STRING_TEMPLATE_CONTENT -&gt; Hello,
+     *                      |--EMBEDDED_EXPRESSION_BEGIN -&gt; \{
```

Comment:
I myself think this should be more nested over flat like this. While we don't do it like these everywhere, we do for block comments and object blocks. I think it makes it clearer the start and end.

I saw the discussion about this before this PR.

Reply:
Can you throw together an example so we can be on the same page?

Reply:
Copied from the original discussion:
````
        |   |           `--STRING_TEMPLATE_BEGIN -> "x\{ [12:26]
        |   |               |--EMBEDDED_EXPRESSION -> EMBEDDED_EXPRESSION [12:31]
        |   |               |   `--IDENT -> x [12:31]
        |   |               `--STRING_TEMPLATE_END -> }x" [12:33]
````

Taken from other ASTs:
````
        |  |  |--BLOCK_COMMENT_BEGIN -> /* [2:0]
        |  |  |  |--COMMENT_CONTENT ->  [2:2]
        |  |  |  `--BLOCK_COMMENT_END -> */ [2:2]

        |  |  |--BLOCK_COMMENT_BEGIN -> /* [3:0]
        |  |  |  |--COMMENT_CONTENT -> *  [3:2]
        |  |  |  |  `--JAVADOC -> JAVADOC [3:3]
        |  |  |  |      |--TEXT ->  [3:3]
        |  |  |  |      `--EOF -> <EOF> [3:4]
        |  |  |  `--BLOCK_COMMENT_END -> */ [3:3]

        |  `--SLIST -> { [4:18]
        |      `--RCURLY -> } [5:4]
````

I am not really asking for you to change your PR if this is the way the group wants to go. I just think the other way makes the hierarchy clearer. There are other areas in our code that flattens it like this does, but like I said, I disagree with that layout. A clearer start/end, makes coding to handle the structure in a custom check easier. The difference between looping until `null` and looping until token type XYZ. I had some custom check have to do this kind of looping, and i thought this structure just complicated the coding.

Example:
````
        |--METHOD_DEF -> METHOD_DEF [4:4]
        |  |--IDENT -> method [4:9]
        |  |--LPAREN -> ( [4:15]
        |  |--PARAMETERS -> PARAMETERS [4:16]
        |  |--RPAREN -> ) [4:16]
````

Reply:
It is debatable for sure; my thought was that now all the content tokens are the same, instead of having beginning/middle/end that all have different delimiters. From a check update point of view, we can pretty much just add the STRING_TEMPLATE_CONTENT token and be done with it.